### ISSUE TITLE
Typescript declaration files

### ISSUE THREAD

**Issue by Robinfr** [parcel-bundler/parcel#1240]
Currently the TypeScript declaration files are not generated even when enabled in the `tsconfig.json`, would be nice if this could also be generated when building.

**Comment by tomtobac** [comment]
Any updates?

**Comment by fathyb** [comment]
Blocked by #465, no type-checking = no declarations. You can open an issue in the TypeScript plugin repo or wait for the plugin to be integrated in Parcel's core.

**Comment by mohsen1** [comment]
You can use tsc to generate type declarations independent of Parcel. Since Parcel is not doing the type-checking it's good to run tsc anyways.

**Comment by vincentfiestada** [comment]
Is it possible to generate type definitions for projects that use both Typescript and ReactJS? Using `tsc` on its own doesn't work, since the TSX needs to be parsed as well. Normally, I would use webpack, but then what would be the point of Parcel?
So, is there a way for the files to go through the normal Parcel pipeline (Babel, etc.) before going through the Typescript compiler?

**Comment by vincentfiestada** [comment]
It looks like the previous errors were actually just legitimate errors being emitted by the compiler. They were being suppressed by parcel. 

I ended up running parcel and then running the Typescript compiler to get declaration files.

**Comment by laosb** [comment]
Can't believe not yet implemented after 1 year and a half.

**Comment by DeMoorJasper** [comment]
@laosb you can just run tsc for typechecking and typedefinitions...

Parcel 2 supports typechecking, not entirely sure how type definition generation in a bundler works as I've never needed that before

**Comment by Xiphe** [comment]
any updates here?

Would be great if we could use Parcel 2 to bundle node packages and include type declarations.

**Comment by DeMoorJasper** [comment]
This is already part of Parcel 2

**Comment by klemenoslaj** [comment]
> This is already part of Parcel 2

I am just trying out Parcel 2. I configured my `tsconfig.json` to generate declarations.
Worth mentioning I am using latest nightly release of parcel with tsc transformer (parcel 2 to my big surprise uses babel typescript as a default).

Is your statement still valid?

**Comment by mischnic** [comment]
Something like this will generate the types at `types.d.ts`:
```js
{
    "main": "dist/index.js"
    "types": "dist/index.d.ts"
	"scripts": {
        "build": "parcel build src/index.js"
	}
}
```

**Comment by klemenoslaj** [comment]
@mischnic you mean just having `types` in package.json should be enough?

I did try it tho, but am not getting it still 🤔

**Comment by mischnic** [comment]
> @mischnic you mean just having types in package.json should be enough?

Yes. 

> I did try it tho, but am not getting it still 🤔

can you open a new issue with a code sample?

**Comment by klemenoslaj** [comment]
> can you open a new issue with a code sample?

Will do :+1:

Trying for the second time I get:

```
@parcel/namer-default: Target "types" declares an output file path of "dist/index.d.ts" which does not match the compiled bundle type "js".
```

**Comment by mischnic** [comment]
> @parcel/namer-default: Target "types" declares an output file path of "dist/index.d.ts" which does not match the compiled bundle type "js".

I have seen this being reported but wasn't able to reproduce it myself...

**Comment by dsedmak** [comment]
Seems like this is still an issue. I can reproduce this reliably on a fresh project.

If we do not have a .parcelrc file or have one that simply extends the default config adding a types entry to package.json causes the declarations to be generated.

With the following .parcelrc:
``` json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{ts,tsx}": ["@parcel/transformer-babel"]
  }
}
```
We get the error: `@parcel/namer-default: Target "main" declares an output file path of "dist/index.js" which does not match the compiled bundle type "ts".`

With the following .parcelrc:
``` json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.{ts,tsx}": ["@parcel/transformer-typescript-tsc"]
  }
}
```
We get the error: `@parcel/namer-default: Target "types" declares an output file path of "dist/index.d.ts" which does not match the compiled bundle type "js".`

**Comment by mischnic** [comment]
> "*.{ts,tsx}": ["@parcel/transformer-babel"]

You're missing `@parcel/transformer-js`. (And that configuration isn't needed since this is what the default config does).

>  Target "types" declares an output file path of "dist/index.d.ts" which does not match the compiled bundle type "js"

Please see https://github.com/parcel-bundler/parcel/issues/4666. Looking into it now...
